File: src/services/api/web-reader.ts
/**
 * 网页阅读器 API 模块
 */
import { invoke } from "@tauri-apps/api/core";
import type { WebSnapshot } from "@/types";

/**
 * 网页抓取结果（完整内容）
 */
export interface FetchResult {
  title: string;
  author?: string;
  siteName?: string;
  content: string;        // 清洗后的 HTML
  textContent: string;    // 纯文本
  excerpt?: string;
  wordCount: number;
}

/**
 * 网页元数据（用于表单自动填充）
 */
export interface WebpageMetadata {
  title: string;
  author?: string;
  siteName?: string;
  description?: string;
  favicon?: string;
}

/**
 * 抓取并清洗网页（完整内容）
 */
export async function fetchWebpage(url: string): Promise<FetchResult> {
  return await invoke<FetchResult>("fetch_webpage", { url });
}

/**
 * 快速获取网页元数据（用于表单自动填充）
 */
export async function fetchWebpageMetadata(url: string): Promise<WebpageMetadata> {
  return await invoke<WebpageMetadata>("fetch_webpage_metadata", { url });
}

/**
 * 保存网页快照
 */
export async function saveSnapshot(
  sourceId: string,
  url: string,
  fetchResult: FetchResult
): Promise<WebSnapshot> {
  return await invoke<WebSnapshot>("save_web_snapshot", {
    sourceId,
    url,
    fetchResult,
  });
}

/**
 * 获取网页快照
 */
export async function getSnapshot(sourceId: string): Promise<WebSnapshot | null> {
  return await invoke<WebSnapshot | null>("get_web_snapshot", { sourceId });
}

/**
 * 将 HTML 转换为 Markdown
 */
export async function convertToMarkdown(html: string): Promise<string> {
  return await invoke<string>("convert_to_markdown", { html });
}

File: src-tauri/src/commands/web_reader.rs
//! 网页阅读器相关命令

use crate::state::AppState;
use crate::web_reader::{FetchResult, WebSnapshot, WebpageMetadata};
use tauri::State;

/// 抓取并清洗网页（完整内容）
#[tauri::command]
pub fn fetch_webpage(state: State<AppState>, url: String) -> Result<FetchResult, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services.web_reader.fetch_webpage(&url)
}

/// 快速获取网页元数据（用于表单自动填充）
#[tauri::command]
pub fn fetch_webpage_metadata(state: State<AppState>, url: String) -> Result<WebpageMetadata, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services.web_reader.fetch_metadata(&url)
}

/// 保存网页快照
#[tauri::command]
pub async fn save_web_snapshot(
    state: State<'_, AppState>,
    source_id: String,
    url: String,
    fetch_result: FetchResult,
) -> Result<WebSnapshot, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services
        .web_reader
        .save_snapshot(&source_id, &url, fetch_result)
        .await
}

/// 获取网页快照
#[tauri::command]
pub async fn get_web_snapshot(state: State<'_, AppState>, source_id: String) -> Result<Option<WebSnapshot>, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services
        .web_reader
        .get_snapshot(&source_id)
        .await
}

/// 将网页内容转换为 Markdown
#[tauri::command]
pub fn convert_to_markdown(state: State<AppState>, html: String) -> Result<String, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    Ok(services.web_reader.convert_to_markdown(&html))
}

